core/jvm/src/main/scala/scalaz/zio/RTS.scala

```diff
@@ -459,8 +460,10 @@ private object RTS {
                                 if (curIo eq null) {
                                   result = value
                                 }
-                              case ExitResult.Terminated(ts) =>
-                                curIo = IO.terminate0(ts)
+                              case ExitResult.Interrupted(e, ts) =>
+                                curIo = IO.terminate0(Errors.TerminatedFiber(e), ts)
```

Comment:
The translation of this `Interrupted` into `IO.terminate` concerns me because we are losing information and I think there will be strange edge cases.

I think you should instead say something like `curIo = interrupt(e, ts)`. It can't exactly be `interrupt` because ordinary `interrupt` method on `FiberContext` will block until interrupted (it's not intended to be called by a fiber on itself).

But it should be possible to write a `selfInterrupt` method that will call `kill0` directly, and modify `kill0` so that adding a callback is optional.

As a short-term fix, you could also do:

```scala
curIo = interrupt(...).fork
```

However, you still have the problem that `interrupt` is not designed to take defects, so you have to add those defects to those being collected in the fiber state. /cc @regiskuckaertz 

Also I suggest to use a plural naming convention on `e` to be consistent: e.g. `es`.